src/radical/pilot/pmgr/launching/saga.py

```diff
@@ -0,0 +1,268 @@
+
+            # cancel pilots
+            for _, job in self._saga_jobs:
+                job.cancel()
+
+            # close job services
+            for url, js in self._saga_js.items():
+                self._log.debug('close js %s', url)
+                js.close()
+
+
+    # --------------------------------------------------------------------------
+    #
+    def can_launch(self, rcfg, pilot):
+
+        # SAGA can launch all pilots
+        return True
+
+
+    # --------------------------------------------------------------------------
+    #
+    def launch_pilots(self, rcfg, pilots):
+
+        js_ep  = rcfg['job_manager_endpoint']
+        with self._saga_lock:
+            if js_ep in self._saga_js:
+                js = self._saga_js[js_ep]
+            else:
+                js = rs.job.Service(js_ep, session=self._saga_session)
+                self._saga_js[js_ep] = js
+
+        # now that the scripts are in place and configured,
+        # we can launch the agent
+        jc = rs.job.Container()
+
+        for pilot in pilots:
+
+            jd_dict = pilot['jd_dict']
+
+            saga_jd_supplement = dict()
+            if 'saga_jd_supplement' in jd_dict:
+                saga_jd_supplement = jd_dict['saga_jd_supplement']
+                del(jd_dict['saga_jd_supplement'])
+
+
+            jd = rs.job.Description()
+            for k, v in jd_dict.items():
+                jd.set_attribute(k, v)
+
+            # set saga_jd_supplement if not already set
+            for key, val in saga_jd_supplement.items():
+                if not jd[key]:
+                    self._log.debug('supplement %s: %s', key, val)
+                    jd[key] = val
+
+            # set saga job description attribute based on env variable(s)
+            # FIXME: is this used?
+            if os.environ.get('RADICAL_SAGA_SMT'):
+                try:
+                    jd.system_architecture['smt'] = \
+                        int(os.environ['RADICAL_SAGA_SMT'])
+                except Exception as e:
+                    self._log.debug('SAGA SMT not set: %s' % e)
+
+          # self._log.debug('jd: %s', pprint.pformat(jd.as_dict()))
+            jc.add(js.create_job(jd))
+
+        jc.run()
+
+        # Order of tasks in `rs.job.Container().tasks` is not changing over the
+        # time, thus it's able to iterate over it and other list(s) all together
+        for j, pilot in zip(jc.get_tasks(), pilots):
+
+            # do a quick error check
+            if j.state == rs.FAILED:
+                self._log.error('%s: %s : %s : %s', j.id, j.state, j.stderr, j.stdout)
+                raise RuntimeError("SAGA Job state is FAILED. (%s)" % j.name)
+
+            pid = pilot['uid']
+
+            # FIXME: update the right pilot
+            with self._saga_lock:
+                self._saga_jobs[pid] = j
+
+
+    # --------------------------------------------------------------------------
+    #
+    def cancel_pilots(self, pids):
```

Comment:
this method is not called in base module

Reply:
Ah, thanks - underscore was missing...

Reply:
I actually meant that there is no calls `launcher._cancel_pilots` (use an update method name here), where `launcher` is an object from `self._launchers` in base module

Reply:
Ah, stupid me - cancel was indeed incomplete.  I fixed that now, but that needed a number of changes actually (need to remember what launcher submitted a pilot so that we can use the same launcher for the cancel call).  You may want to directly look at this commit: 71d3379d6